Switch from unmaintained pytube to yt-dlp. Check user is subscribed to YTM premium

tronikos opened the pull request:
pytube hasn't been updated for years. Music Assistant is actually using a 1 year old fork by the ytmusic maintainer.
On the contrary, yt-dlp development is active. I found it more reliable but a few seconds slower.

marcelveldt commented:
If this is more reliable, why not start using it as the default ?
We can then maybe just tru to improve the speed later. I think folks rather have it reliable than fast and currently you see many issues popping up that playback is failing.

My opinion would be to make this switch the default (but I'm not using YTM) so I'll leave the decision to @MarvinSchenkel

tronikos commented:
It's very slow. About 20-30 seconds for each song. Unless we can make it faster we can't make it the default.

MarvinSchenkel commented:
> It's very slow. About 20-30 seconds for each song. Unless we can make it faster we can't make it the default.

I will have a look. Smells like it is trying to download the entire song before returning any data. We might be lucky if we can make it `yield` bytes as soon as they are downloaded from YTM.

tronikos commented:
Note I'm calling `extract_info` with `download=False` so it doesn't download the audio bytes. It downloads the webpage, JSON, and m3u8 to give you streaming URLs.

MarvinSchenkel commented:
Ah that makes sense! Did you figure out which part takes a long time? Is it the `ydl.extract_info()` function?

tronikos commented:
Yes. Specifically the inner `YoutubeIE._real_extract`

tronikos commented:
It seems the slowness is because I was running Music Assistant in the VS Code debugger from a dev container in Windows. 

Running the following code directly it's only about 2 to 5 seconds:

```py
import yt_dlp

def _extract_stream_url(item_id):
    ydl = yt_dlp.YoutubeDL()
    info = ydl.extract_info(f"https://www.youtube.com/embed/{item_id}", download=False)
    audio_only_formats = [
        f
        for f in info["formats"]
        if f.get("audio_ext", "none") == "m4a" and f.get("video_ext", "none") == "none"
    ]
    # Formats are sorted from worse to better quality. Use the last/best one.
    return audio_only_formats[-1]["url"]

print(_extract_stream_url("5GD4hsz9W78"))
```

So no need to use a subprocess. Using the CLI I don't think there is a way to get the streaming URL anyway.

Do you want to fully switch to yt-dlp? Or keep pytube around with an option to allow users to switch to that?

Regarding premium vs free I'll let you handle it. FWIW, I just tried the latest released version of Music Assistant on an account not subscribed to YTM Premium and surprisingly I could still play all the songs I tried.

MarvinSchenkel commented:
> It seems the slowness is because I was running Music Assistant in the VS Code debugger from a dev container in Windows.
> 
> Running the following code directly it's only about 2 to 5 seconds:
> 
> ```python
> import yt_dlp
> 
> def _extract_stream_url(item_id):
>     ydl = yt_dlp.YoutubeDL()
>     info = ydl.extract_info(f"https://www.youtube.com/embed/{item_id}", download=False)
>     audio_only_formats = [
>         f
>         for f in info["formats"]
>         if f.get("audio_ext", "none") == "m4a" and f.get("video_ext", "none") == "none"
>     ]
>     # Formats are sorted from worse to better quality. Use the last/best one.
>     return audio_only_formats[-1]["url"]
> 
> print(_extract_stream_url("5GD4hsz9W78"))
> ```
> 
> So no need to use a subprocess. Using the CLI I don't think there is a way to get the streaming URL anyway.
> 
> Do you want to fully switch to yt-dlp? Or keep pytube around with an option to allow users to switch to that?
> 
> Regarding premium vs free I'll let you handle it. FWIW, I just tried the latest released version of Music Assistant on an account not subscribed to YTM Premium and surprisingly I could still play all the songs I tried.

Great find! Let's fully switch to yt-dlp 🚀 . In the mean time, I will figure out a way to actively block people without a YTM premium subscription.

One minor thing, instead of the custom code to figure out the audio stream, can we leverage the [format 'bestaudio' from yt-dlp](https://github.com/yt-dlp/yt-dlp?tab=readme-ov-file#format-selection)?

tronikos commented:
I switched to using a format_selector with bestaudio.

But I just realized a deal breaker. The streams from yt-dlp aren't high quality. They are at 128 Kbps instead of 256 Kbps. So we can definitely not switch to it. Up to you if you want to close this PR or offer the advanced option to allow users switching to yt-dlp. I added a warning to the description of the setting that it only supports 128 Kbps.

tronikos commented on music_assistant/server/providers/ytmusic/__init__.py:
```diff
@@ -826,15 +827,13 @@ async def _extract_stream_url_using_yt_dlp(self, item_id: str) -> str:
+            with yt_dlp.YoutubeDL(ydl_opts) as ydl:
+                url = f"https://www.youtube.com/embed/{item_id}"
+                info = ydl.extract_info(url, download=False)
+                format_selector = ydl.build_format_selector("m4a/bestaudio")
```
Unfortunately not. It only applies if you call the download function.

MarvinSchenkel commented:
> I switched to using a format_selector with bestaudio.
> 
> But I just realized a deal breaker. The streams from yt-dlp aren't high quality. They are at 128 Kbps instead of 256 Kbps. So we can definitely not switch to it. Up to you if you want to close this PR or offer the advanced option to allow users switching to yt-dlp. I added a warning to the description of the setting that it only supports 128 Kbps.

I am not 100% sure if the current setup is also grabbing 256kbps streams, since it also grabs streams from `youtube.com/embed`. Did you check that by any chance?

tronikos commented:
Yes I confirmed with the current setup I get 256kbps for songs that support it.

MarvinSchenkel commented:
> Yes I confirmed with the current setup, i.e. pytube, I get 256kbps for songs that support it.

Do you still get 128kbps streams when you remove your 'm4a/bestaudio' filter and look for other audio streams?

tronikos commented:
With `ydl.list_formats(info)` I get:

```
ID  EXT   RESOLUTION FPS CH │   FILESIZE  TBR PROTO │ VCODEC         VBR ACODEC      ABR ASR MORE INFO 
──────────────────────────────────────────────────────────────────────────────────────────────────────────────
sb3 mhtml 48x27        0    │                 mhtml │ images                                 storyboard
sb2 mhtml 45x45        0    │                 mhtml │ images                                 storyboard
sb1 mhtml 90x90        0    │                 mhtml │ images                                 storyboard
sb0 mhtml 180x180      0    │                 mhtml │ images                                 storyboard
233 mp4   audio only        │                 m3u8  │ audio only         unknown             Default
234 mp4   audio only        │                 m3u8  │ audio only         unknown             Default
139 m4a   audio only      2 │    2.25MiB  49k https │ audio only         mp4a.40.5   49k 22k low, m4a_dash
249 webm  audio only      2 │    2.32MiB  50k https │ audio only         opus        50k 48k low, webm_dash
250 webm  audio only      2 │    3.09MiB  67k https │ audio only         opus        67k 48k low, webm_dash
140 m4a   audio only      2 │    5.98MiB 130k https │ audio only         mp4a.40.2  130k 44k medium, m4a_dash
251 webm  audio only      2 │    6.14MiB 133k https │ audio only         opus       133k 48k medium, webm_dash
602 mp4   144x144     13    │ ~  2.81MiB  61k m3u8  │ vp09.00.10.08  61k video only
269 mp4   144x144     25    │ ~  2.94MiB  64k m3u8  │ avc1.4D400B    64k video only
160 mp4   144x144     25    │  356.94KiB   8k https │ avc1.4D400B     8k video only          144p, mp4_dash
603 mp4   144x144     25    │ ~  2.97MiB  64k m3u8  │ vp09.00.11.08  64k video only
278 webm  144x144     25    │  402.80KiB   9k https │ vp9             9k video only          144p, webm_dash
229 mp4   240x240     25    │ ~  3.11MiB  67k m3u8  │ avc1.4D400C    67k video only
133 mp4   240x240     25    │  457.16KiB  10k https │ avc1.4D400C    10k video only          240p, mp4_dash
604 mp4   240x240     25    │ ~  2.92MiB  63k m3u8  │ vp09.00.20.08  63k video only
242 webm  240x240     25    │  367.77KiB   8k https │ vp9             8k video only          240p, webm_dash
230 mp4   360x360     25    │ ~  7.15MiB 155k m3u8  │ avc1.4D4015   155k video only
134 mp4   360x360     25    │  599.20KiB  13k https │ avc1.4D4015    13k video only          360p, mp4_dash
18  mp4   360x360     25  2 │ ≈  6.52MiB 141k https │ avc1.42001E        mp4a.40.2       44k 360p
605 mp4   360x360     25    │ ~  6.92MiB 150k m3u8  │ vp09.00.21.08 150k video only
243 webm  360x360     25    │  443.22KiB   9k https │ vp9             9k video only          360p, webm_dash
231 mp4   480x480     25    │ ~  7.56MiB 164k m3u8  │ avc1.4D401E   164k video only
135 mp4   480x480     25    │  882.62KiB  19k https │ avc1.4D401E    19k video only          480p, mp4_dash
606 mp4   480x480     25    │ ~  6.86MiB 149k m3u8  │ vp09.00.30.08 149k video only
244 webm  480x480     25    │  378.48KiB   8k https │ vp9             8k video only          480p, webm_dash
232 mp4   720x720     25    │ ~  8.49MiB 184k m3u8  │ avc1.64001F   184k video only
136 mp4   720x720     25    │    1.39MiB  30k https │ avc1.64001F    30k video only          720p, mp4_dash
609 mp4   720x720     25    │ ~  7.24MiB 157k m3u8  │ vp09.00.31.08 157k video only
247 webm  720x720     25    │  681.07KiB  14k https │ vp9            14k video only          720p, webm_dash
270 mp4   1080x1080   25    │ ~ 10.30MiB 223k m3u8  │ avc1.640020   223k video only
137 mp4   1080x1080   25    │    2.52MiB  55k https │ avc1.640020    55k video only          1080p, mp4_dash
614 mp4   1080x1080   25    │ ~  8.67MiB 188k m3u8  │ vp09.00.40.08 188k video only
248 webm  1080x1080   25    │    1.65MiB  36k https │ vp9            36k video only          1080p, webm_dash
```

The best mp4a is format id 140 (the one selected). The best opus is format id 251. Both are a bit higher than 128kbps.

Ideally we want the format id 141 which is 256kbps but that one is only available to YTM premium. If we had cookies we could pass them to yt-dlp https://github.com/ytdl-org/youtube-dl/#how-do-i-pass-cookies-to-youtube-dl

MarvinSchenkel commented:
Well we do have an access token, can we somehow transform that into the cookie info required by yt-dlp? This would kill two birds with one stone, because then we can just filter to use YTM premium streams, practically disallowing free usage here.

MarvinSchenkel commented:
> I don't see how to transform the access token to cookies.

We could try to find the python version of the `--add-header` flag and pass `"Authorization": f'Bearer {tokeninfo["accessToken"]}'`

MarvinSchenkel commented:
Been playing around a bit, but no luck yet. What I tried so far:

1. Adding custom `http_headers` to `params`:
```python
ydl_opts = {
   "http_headers": self._headers
  }
  with yt_dlp.YoutubeDL(ydl_opts) as ydl:
  .....
```
This results in a 401 error.

2. Using an [oauth2 plugin for yt-dlp](https://github.com/coletdjnz/yt-dlp-youtube-oauth2/tree/master). I can see the requests to youtube.com/embed get routed through the plugin. However, I still only see low and medium streams in the `ydl.list_formats(info)`. @tronikos maybe you have better luck with this route

MarvinSchenkel commented:
Cool, I sorted it out. When I login via the oauth plugin AND use a different end-point (`url = f"https://music.youtube.com/watch?v={item_id}"`), I am able to find the infamous stream with id 141:
```
ID  EXT   RESOLUTION FPS CH │   FILESIZE  TBR PROTO │ VCODEC         VBR ACODEC      ABR ASR MORE INFO
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────
sb2 mhtml 48x27        0    │                 mhtml │ images                                 storyboard
sb1 mhtml 45x45        0    │                 mhtml │ images                                 storyboard
sb0 mhtml 90x90        0    │                 mhtml │ images                                 storyboard
233 mp4   audio only        │                 m3u8  │ audio only         unknown             Default, IOS
234 mp4   audio only        │                 m3u8  │ audio only         unknown             Default, IOS
139 m4a   audio only      2 │    2.61MiB  49k https │ audio only         mp4a.40.5   49k 22k low, IOS, m4a_dash
249 webm  audio only      2 │    2.64MiB  49k https │ audio only         opus        49k 48k low, WEB, webm_dash
250 webm  audio only      2 │    3.48MiB  65k https │ audio only         opus        65k 48k low, WEB, webm_dash
140 m4a   audio only      2 │    6.92MiB 130k https │ audio only         mp4a.40.2  130k 44k medium, IOS, m4a_dash
251 webm  audio only      2 │    6.93MiB 130k https │ audio only         opus       130k 48k medium, WEB, webm_dash
141 m4a   audio only      2 │   13.77MiB 258k https │ audio only         mp4a.40.2  258k 44k high, IOS-M, m4a_dash
774 webm  audio only      2 │   13.79MiB 258k https │ audio only         opus       258k 48k high, WEB-M, webm_dash
269 mp4   144x144     25    │ ~  3.85MiB  72k m3u8  │ avc1.4D400B    72k video only          IOS
160 mp4   144x144     25    │  756.74KiB  14k https │ avc1.4D400B    14k video only          144p, IOS, mp4_dash
603 mp4   144x144     25    │ ~  4.07MiB  76k m3u8  │ vp09.00.11.08  76k video only          IOS
278 webm  144x144     25    │  882.93KiB  16k https │ vp9            16k video only          144p, WEB, webm_dash
229 mp4   240x240     25    │ ~  4.85MiB  91k m3u8  │ avc1.4D400C    91k video only          IOS
133 mp4   240x240     25    │    1.31MiB  25k https │ avc1.4D400C    25k video only          240p, IOS, mp4_dash
604 mp4   240x240     25    │ ~  4.88MiB  91k m3u8  │ vp09.00.20.08  91k video only          IOS
242 webm  240x240     25    │    1.30MiB  24k https │ vp9            24k video only          240p, WEB, webm_dash
230 mp4   360x360     25    │ ~ 10.91MiB 204k m3u8  │ avc1.4D4015   204k video only          IOS
134 mp4   360x360     25    │    2.38MiB  45k https │ avc1.4D4015    45k video only          360p, IOS, mp4_dash
18  mp4   360x360     25  2 │ ≈  9.26MiB 173k https │ avc1.42001E        mp4a.40.2       44k 360p, WEB
605 mp4   360x360     25    │ ~ 10.67MiB 200k m3u8  │ vp09.00.21.08 200k video only          IOS
243 webm  360x360     25    │    2.09MiB  39k https │ vp9            39k video only          360p, WEB, webm_dash
231 mp4   480x480     25    │ ~ 13.83MiB 259k m3u8  │ avc1.4D401E   259k video only          IOS
135 mp4   480x480     25    │    4.03MiB  75k https │ avc1.4D401E    75k video only          480p, IOS, mp4_dash
606 mp4   480x480     25    │ ~ 12.37MiB 232k m3u8  │ vp09.00.30.08 232k video only          IOS
244 webm  480x480     25    │    3.05MiB  57k https │ vp9            57k video only          480p, WEB, webm_dash
232 mp4   720x720     25    │ ~ 21.95MiB 411k m3u8  │ avc1.4D401F   411k video only          IOS
136 mp4   720x720     25    │    8.36MiB 156k https │ avc1.4D401F   156k video only          720p, IOS, mp4_dash
609 mp4   720x720     25    │ ~ 16.76MiB 314k m3u8  │ vp09.00.31.08 314k video only          IOS
247 webm  720x720     25    │    5.53MiB 103k https │ vp9           103k video only          720p, WEB, webm_dash
270 mp4   1080x1080   25    │ ~ 34.52MiB 646k m3u8  │ avc1.640020   646k video only          IOS
137 mp4   1080x1080   25    │   15.50MiB 290k https │ avc1.640020   290k video only          1080p, IOS, mp4_dash
614 mp4   1080x1080   25    │ ~ 26.66MiB 499k m3u8  │ vp09.00.40.08 499k video only          IOS
248 webm  1080x1080   25    │   12.18MiB 228k https │ vp9           228k video only          1080p, WEB, webm_dash
```

All that's left now is figuring out if we can somehow pass the access token we already in MA down to the oauth2 plugin, since it is now using a second oauth process.

MarvinSchenkel commented:
All done! Also added a check whether a user is a premium user. @tronikos can you dun this locally and see if this is all working fine for you before we merge it to `main`? You can do an editable install to workaround the delay of ytdlp and start MA from the commandline:
```
pip install -e ".[server]"
mass
```

tronikos commented:
Sorry for the delay. Your new code looks good. I only fixed a couple of small things. I''ll test and report back within a day or two.

Great find of https://github.com/coletdjnz/yt-dlp-youtube-oauth2 I don't quite like that it's experimental and that it has a warning that it might break at any time. But as long as you are aware of it and you are fine with it I have no objections.

MarvinSchenkel commented on music_assistant/server/providers/ytmusic/__init__.py:
```diff
@@ -813,7 +828,7 @@ async def _user_has_ytm_premium(self) -> bool:
         stream_format = await self._get_stream_format(YTM_PREMIUM_CHECK_TRACK_ID)
         # Only premium users can stream the HQ stream of this song
-        return stream_format["format"] == "141 - audio only (high)"
+        return stream_format["format_id"] == "141"
```
Are you sure this is formatted as a string?

MarvinSchenkel commented:
> Sorry for the delay. Your new code looks good. I only fixed a couple of small things. I''ll test and report back within a day or two.
> 
> Great find of https://github.com/coletdjnz/yt-dlp-youtube-oauth2 I don't quite like that it's experimental and that it has a warning that it might break at any time. But as long as you are aware of it and you are fine with it I have no objections.

100% agree. Should it break, I am planning on writing my own plugin for yt dlp that just accepts an access token to authenticate requests and push it to pip.

MarvinSchenkel commented:
Merged! Thanks for this awesome suggestion @tronikos

SuperSandro2000 commented:
Just wanted to let everyone know that I enjoyed listening to https://www.youtube.com/watch?v=dQw4w9WgXcQ
